### PR TITLE
Infograph 

### PR DESCRIPTION
**Pull request by tonydavis629**
# Pull Request Template

## Description

This PR implements Infograph as a ModularTorchModel. Infograph is a semi-supervised graph convolutional neural network.

## Type of change

Please check the option that is related to your PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/tests/test_infograph.py
Comment (rbharath) [inline deepchem/models/tests/test_infograph.py:23]: This will load the full dataset which may be slow. Perhaps better to make a tiny dataset of 20 compounds directly for testing (the data_points specification below only avoids training on the full set I think and doesn't avoid the load)

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:10]: Perhaps InfographEncoder since we will have many encoder layers

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:53]: Can we reuse one of the existing feed forward nn.modules? I think we have a few

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:103]: I think this is only the fine tuning use case? Can you add an example for the semisupervised pretraining?

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:302]: All these different values for measure should be documented in the doctoring for the class

File: docs/source/api_reference/models.rst
Comment (rbharath) [inline docs/source/api_reference/models.rst:396]: All the layers should be in the docs. We're trying to build up and document DeepChem's collection of layers so users can start to reuse layers

File: deepchem/models/tests/test_infograph.py
Comment (tonydavis629) [reply deepchem/models/tests/test_infograph.py:23]: I can change it, but these functions are used in about a dozen tests. Maybe would be better to a create a load function specifically for testing.

File: deepchem/models/torch_models/infograph.py
Comment (tonydavis629) [reply deepchem/models/torch_models/infograph.py:53]: It is a feedforward with a skip connection, which does not match any of the current FF layers

File: deepchem/models/torch_models/infograph.py
Comment (tonydavis629) [reply deepchem/models/torch_models/infograph.py:103]: This is the semi-supervised training. There is no infograph training loop that is strictly supervised. There is only semi-supervised and unsupervised.

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:53]: Minor note, FFSkip would be better since we don't use snake case in class names

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [reply deepchem/models/torch_models/infograph.py:108]: Ah I see. Can you add an unsupervised example as well and add a discussion to the doctoring about the two modes of training and how they differ? Lots of detail is good here since this is the first model of its type in deepchem

File: docs/source/api_reference/models.rst
Comment (rbharath) [inline docs/source/api_reference/models.rst:402]: We want these to go with the other layers since they aren't models and instead are nn.layers

File: deepchem/models/tests/test_infograph.py
Comment (rbharath) [inline deepchem/models/tests/test_infograph.py:9]: Let's split this up into two simpler helper functions, one for classification and one for regression? A good rule of thumb is that a unit test shouldn't take thought to read through. This helper is complex enough it requires additional thought

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:40]: Can you add a docstring here and document the expected shape of `data`?

File: deepchem/models/tests/test_infograph.py
Comment (rbharath) [inline deepchem/models/tests/test_infograph.py:78]: Can we add a unit test for the unsupervised pretraining?

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:116]: Do we need `kwargs`? In general better not to use `kwargs` since that introduces a bigger possibility for bugs

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:161]: This should have a docstring

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:175]: I think this is a private helper so should be `_unsup_loss`

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:112]: I'm feeling so-so about the use of a flag to gate the unsupervised loss. Do we want to establish that as the defacto pattern for how self-supervised models turn pretraining on/off? Let's discuss offline before we make the final call

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:217]: Do you want to do `_local_global_loss` to mark this as private? I think the underscore should come up front

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:365]: Can you add a usage example?

File: deepchem/models/torch_models/layers.py
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:27]: I'm assuming these changes will be rebased out once the MLP PR is merged?

File: deepchem/models/torch_models/infograph.py
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:251]: Should be `_global_global_loss`?

**Review by rbharath — COMMENTED** [review]
Almost at the finish line here! Some improvements needed on testing and documentation and a rebase after the merge in the MLP layer PR but I think one more pass and we can get this in

File: deepchem/models/torch_models/infograph.py
Comment (tonydavis629) [reply deepchem/models/torch_models/infograph.py:116]: It is necessary for all torch_models so they can access the optional torch_model arguments (like batch_size, tensorboard, etc)

File: deepchem/models/torch_models/infograph.py
Comment (tonydavis629) [reply deepchem/models/torch_models/infograph.py:112]: Infograph is written with different types of training already in mind, whereas I think many models are made with only one type of training task. Modifying the loss_func is how I intend the training task to be modified in ModularTorchModel. I think in the case of infograph it makes sense, but maybe it should not be the usual pattern.

File: docs/source/api_reference/models.rst
Comment (rbharath) [reply docs/source/api_reference/models.rst:402]: Quick reminder on this TODO

**Review by rbharath — APPROVED** [review]
LGTM. One last minor comment about moving the layer documentation to `layers.rst` but clear to merge after once CI is good

**Pull request merged**
